File: Pi-Dev22K24-houssem/codemastersfrontend/src/app/models/post.ts
import { Comment } from "./comment";
import { Tag } from "./tag";
import { User } from "./user";

// export interface Post {
//   id: number;
//   title: string;
//   content: string;
//   // image: string;
//   image?: Blob;
//   imageSrc?: string;
//   video: string;
//   creationDate: Date;
//   tags: Tag[];
//   comments: Comment[];
//   likes: number;
//   dislikes: number;
//   likesByUser: User[];
//   dislikesByUser: User[] | null; // Use 'User[] | null' instead of 'User[]'
//   user: User | null;

// }
export interface Post {
  id: number;
  title: string;
  content: string;
  imageId?: number; // Nouvelle propriété pour stocker l'ID de l'image
  image?: Blob;
  imageSrc?: string;
  
  creationDate: Date;
  tags: Tag[];
  comments: Comment[];
  likes: number;
  dislikes: number;
  likesByUser: User[];
  dislikesByUser: User[] | null; // Use 'User[] | null' instead of 'User[]'
  user: User | null;
}

File: Pi-Dev22K24-houssem/codemastersbackend/src/main/java/tn/esprit/codemasters/entity/Post.java
package tn.esprit.codemasters.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Entity
@Getter
@Setter
@ToString
@NoArgsConstructor
@AllArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
@Table(name = "posts")
public class Post {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    Long id;

    String title;
    String content;
    //String image;
    //String video;

    @Lob
    @Column(name = "image")
    private byte[] image;

    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "creation_date")
    Date creationDate;

    @ManyToMany(cascade = CascadeType.ALL)
    @JoinTable(
            name = "post_tags",
            joinColumns = @JoinColumn(name = "post_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    @JsonIgnore
    private List<Tag> tags;

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL)
    @JsonManagedReference
    @JsonIgnore // Ignore the comments during serialization to break the circular reference
    private List<Comment> comments;

    @PrePersist
    protected void onCreate() {
        creationDate = new Date();
    }

    @PostLoad
    protected void onLoad() {
        if (getCreationDate() != null) {
            creationDate = new Date(getCreationDate().getTime());
        }
    }

   /* @Transient
    @JsonIgnore
    private List<User> likesByUser = new ArrayList<>();*/

    @OneToMany
    @JsonIgnore
    List<User> liked = new ArrayList<>();

    public List<User> getliked() {
        return liked;
    }

    public void setliked(List<User> liked) {
        this.liked = liked;
    }

    @ManyToOne
    @JsonIgnore // Ignore the user during serialization to avoid circular reference
    User user;
}